### PR TITLE
Overhaul the validation generator

### PR DESCRIPTION
**Pull request by thockin**
This now works in all sorts of combinations - valiadtions can be declared on types, fields, or both.  Validations can exist for values, or lists, or maps.  You can validate a list in whole or each item in it.  Likewise for maps, and you can validate map keys.

There are a bunch of FIXMEs left, but this seems to work.

### CODE REVIEW

File: pkg/apis/core/v1/zz_generated.validations.go [hunk 1]
```diff
@@ -35,120 +35,4614 @@ func init() { localSchemeBuilder.Register(RegisterValidations) }
 // RegisterValidations adds validation functions to the given scheme.
 // Public to allow building arbitrary schemes.
 func RegisterValidations(scheme *runtime.Scheme) error {
-	scheme.AddValidationFunc(&v1.Pod{}, func(obj, oldObj interface{}, subresources ...string) field.ErrorList {
+	scheme.AddValidationFunc(new(v1.Binding), func(obj, oldObj interface{}, subresources ...string) field.ErrorList {
```
Comment (jpbetz) [inline pkg/apis/core/v1/zz_generated.validations.go:38]: Not to self: This registers validations for all types now, not just root types.

File: staging/src/k8s.io/code-generator/cmd/validation-gen/targets.go [hunk 2]
```diff
@@ -205,47 +200,40 @@ func GetTargets(context *generator.Context, args *Args) []generator.Target {
...
 		// Deterministic ordering helps in logs and debugging.
 		slices.SortFunc(rootTypes, func(a, b *types.Type) int {
 			return cmp.Compare(a.Name.String(), b.Name.String())
 		})
 
-		validationFunctionTypes := map[*types.Type]*callNode{}
+		// Deterministic ordering helps in logs and debugging.
+		slices.SortFunc(rootTypes, func(a, b *types.Type) int {
+			return cmp.Compare(a.Name.String(), b.Name.String())
+		})
```
Comment (jpbetz) [inline staging/src/k8s.io/code-generator/cmd/validation-gen/targets.go:226]: @jpbetz follow up: Remove accidental redundant sort.

File: staging/src/k8s.io/code-generator/cmd/validation-gen/validation.go [hunk 3]
```diff
@@ -180,476 +180,533 @@ func (g *genValidations) Init(c *generator.Context, w io.Writer) error {
...
-		return types.Name{}, false
-	}
-	return types.Name{Package: pkg, Name: "Validate_" + t.Name.Name}, true
+// typeNode carries validation informatiuon for a single type.
+type typeNode struct {
+	underlyingType *types.Type
+	validations    []validators.FunctionGen
+	children       []*childNode // populated when parent is a Struct
+	elem           *childNode   // populated when parent is a list
+	key            *childNode   // populated when parent is a map
```
Comment (jpbetz) [inline staging/src/k8s.io/code-generator/cmd/validation-gen/validation.go:230]: This makes me so happy. The defaulter format was so confusing to code against.

File: staging/src/k8s.io/code-generator/cmd/validation-gen/validation.go [hunk 4]
```diff
@@ -180,476 +180,533 @@ func (g *genValidations) Init(c *generator.Context, w io.Writer) error {
...
+)
 
-	if root {
-		// the root node is always a pointer
-		parent.elem = true
+// FIXME: passing comments here is nasty
+// discover walks the type graph, starting at t, and registers all types into
+// knownTypes.  The specified comments represent the parent context for this
+// type - the type comments for a type definition or the field comments for a
+// field.
```
Comment (jpbetz) [inline staging/src/k8s.io/code-generator/cmd/validation-gen/validation.go:261]: @jpbetz follow up

File: pkg/apis/core/v1/zz_generated.validations.go (hunk 1 above)
Comment (thockin) [reply pkg/apis/core/v1/zz_generated.validations.go:38]: Oh, that was not the intent, that seems like a bug.  I wanted to emit code for all types but only register root types (types with...)

File: staging/src/k8s.io/code-generator/cmd/validation-gen/targets.go (hunk 2 above)
Comment (thockin) [reply staging/src/k8s.io/code-generator/cmd/validation-gen/targets.go:226]: Oh, I must have missed that

Comment (jpbetz) [reply staging/src/k8s.io/code-generator/cmd/validation-gen/targets.go:226]: easy peasy, I'll drop in a fix for a few of these details shortly.

**Pull request merged**